### ISSUE TITLE
Consensus Failure

### ISSUE THREAD

**Issue by satsuta** [cybercongress/cyberd#293]
### Current Behavior
[comment]: # (Describe what actually happened.)
  
Panicked on a Consensus Failure.

After restart container fall with `ERROR: Error during handshake: Error on replay: Wrong Block.Header.AppHash`

### Reproduction
[comment]: # (Describe how we can replicate the bug step by step.)
 "cyberd/cyberd:euler-3"      "cyberd/cyberd@sha256:4d10b9dfc258f5b6f5b58f2ecf5edf945663ff1211836c690a5722a6a03dacfd"  
 i5-6400 CPU @ 2.70GHz, RAM 16GB.  GeForce GTX 1070.  Driver Version: 410.104     CUDA version: 10.0 

### Log
```
I[2019-03-07|03:06:10.079] Executed block                               module=state height=336499 validTxs=0 invalidTxs=0
I[2019-03-07|03:06:10.086] Committed state                              module=state height=336499 txs=0 appHash=7C376C11008177D47C87E8C69EE613F6501685F19DF17DE84D9E1D6777E8954E
I[2019-03-07|03:06:10.620] Executed block                               module=state height=336500 validTxs=0 invalidTxs=0
I[2019-03-07|03:06:10.626] Committed state                              module=state height=336500 txs=0 appHash=7C376C11008177D47C87E8C69EE613F6501685F19DF17DE84D9E1D6777E8954E
E[2019-03-07|03:06:11.115] CONSENSUS FAILURE!!!                         module=consensus err="Panicked on a Consensus Failure: +2/3 committed an invalid block: Wrong Block.Header.AppHash.  Expected 7C376C11008177D47C87E8C69EE613F6501685F19DF17DE84D9E1D6777E8954E, got F77C8A0232244DFEB52C54DB78CF92FD27EF078D99DDE2AD9E112D47F52386A0" stack="goroutine 535874 [running]:\nruntime/debug.Stack(0xc031281018, 0xf1b040, 0xc03ceff2d0)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0xa7\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine.func2(0xc000bff180, 0x1237338)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:611 +0x57\npanic(0xf1b040, 0xc03ceff2d0)\n\t/usr/local/go/src/runtime/panic.go:513 +0x1b9\ngithub.com/tendermint/tendermint/libs/common.PanicConsensus(0xf1b040, 0xc03ceff2c0)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/libs/common/errors.go:239 +0xb8\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).finalizeCommit(0xc000bff180, 0x52275)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:1278 +0x48e\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).tryFinalizeCommit(0xc000bff180, 0x52275)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:1255 +0x451\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).addProposalBlockPart(0xc000bff180, 0xc01a752320, 0xc022f9b3b0, 0x28, 0x0, 0x0, 0x18b23d96d41281a3)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:1494 +0x4e7\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).handleMsg(0xc000bff180, 0x1330f80, 0xc01a752320, 0xc022f9b3b0, 0x28)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:682 +0x7a0\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine(0xc000bff180, 0x0)\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:642 +0x66f\ncreated by github.com/tendermint/tendermint/consensus.(*ConsensusState).OnStart\n\t/root/go/pkg/mod/github.com/tendermint/tendermint@v0.30.1/consensus/state.go:334 +0x13d\n"
E[2019-03-07|03:11:48.530] Connection failed @ recvRoutine (reading byte) module=p2p peer=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 conn=MConn{95.216.160.133:26656} err=EOF
E[2019-03-07|03:11:48.530] Stopping peer for error                      module=p2p peer="Peer{MConn{95.216.160.133:26656} 06326633abf29c1ab8203ff062d858ae0fb3dc60 out}" err=EOF
E[2019-03-07|03:12:03.339] Connection failed @ recvRoutine (reading byte) module=p2p peer=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 conn=MConn{95.216.160.133:26656} err=EOF
E[2019-03-07|03:12:03.339] Stopping peer for error                      module=p2p peer="Peer{MConn{95.216.160.133:26656} 06326633abf29c1ab8203ff062d858ae0fb3dc60 out}" err=EOF
E[2019-03-07|03:12:03.339] MConnection flush failed                     module=p2p peer=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 err="write tcp 172.17.0.2:54846->95.216.160.133:26656: use of closed network connection"
E[2019-03-07|03:20:01.489] Connection failed @ recvRoutine (reading byte) module=p2p peer=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 conn=MConn{95.216.160.133:26656} err=EOF
E[2019-03-07|03:20:01.489] Stopping peer for error                      module=p2p peer="Peer{MConn{95.216.160.133:26656} 06326633abf29c1ab8203ff062d858ae0fb3dc60 out}" err=EOF
E[2019-03-07|03:20:28.280] Dialing failed                               module=pex addr=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 err="auth failure: secrect conn failed: read tcp 172.17.0.2:54850->95.216.160.133:26656: i/o timeout" attempts=0
E[2019-03-07|03:20:58.276] Dialing failed                               module=pex addr=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 err="auth failure: secrect conn failed: read tcp 172.17.0.2:54852->95.216.160.133:26656: i/o timeout" attempts=1
E[2019-03-07|03:21:28.274] Dialing failed                               module=pex addr=06326633abf29c1ab8203ff062d858ae0fb3dc60@95.216.160.133:26656 err="auth failure: secrect conn failed: read tcp 172.17.0.2:54854->95.216.160.133:26656: i/o timeout" attempts=2
E[2019-03-07|03:21:59.864] Dialing failed 
```
and after restart 
```
I[2019-03-07|14:11:39.752] Starting ABCI with Tendermint                module=main 
I[2019-03-07|14:11:42.041] Loading mem state                            module=main 
I[2019-03-07|14:13:02.109] App loaded                                   module=main time=1m20.068646999s
ERROR: Error during handshake: Error on replay: Wrong Block.Header.AppHash.  Expected 7C376C11008177D47C87E8C69EE613F6501685F19DF17DE84D9E1D6777E8954E, got F77C8A0232244DFEB52C54DB78CF92FD27EF078D99DDE2AD9E112D47F52386A0
I[2019-03-07|14:13:09.037] Starting ABCI with Tendermint                module=main 
I[2019-03-07|14:13:11.381] Loading mem state                            module=main 
```

**Comment by hleb-albau** [comment]
Current chain corrupted :( 
Will investigate, how wrong app hash was committed.

**Comment by hleb-albau** [comment]
Current chain corrupted :( 
Will investigate, how wrong app hash was committed.

**Comment by hleb-albau** [comment]
We found the problem. In last versions, Cosmos-sdk **x/bank** module introduced new method **DelegateCoins**, that was not covered in in-memory account balances index. Previously, the default **SetCoins** method was used by module **x/staking**. Right now, if you delegate tokens, the in-memory index is not updated.

**Comment by hleb-albau** [comment]
Fixed in 0.1.4
